### PR TITLE
[RNMobile] Refactor Dark Mode HOC

### PR DESCRIPTION
**Pull request by Tug**
## Description
Addresses feedbacks from https://github.com/WordPress/gutenberg/pull/17228
This PR fixes some antipattern in the naming of the HOC used for dark mode on gutenberg mobile.
List of changes:
- Moved from `@wordpress/components` to `@wordpress/compose`
- Renamed HOC from `withTheme` to `withPreferredColorScheme` and move it to  `packages/compose/higher-order/with-preferred-color-scheme`
- Renamed injected prop to compute styles from `useStyle` to `getStylesFromColorScheme`
- Add new hook `usePreferredColorScheme` in `packages/compose/hooks/use-preferred-color-scheme` that simply [exports](https://github.com/WordPress/gutenberg/pull/17552/files#diff-a028bc66e9dad38eeac7ce1197477dbdR11) `useDarkModeContext ` from [`react-native-dark-mode`](https://www.npmjs.com/package/react-native-dark-mode)
- Change [implementation](https://github.com/WordPress/gutenberg/pull/17552/files#diff-6a8479ece0ad578559d72f5f58e2bdf4R7) of the `withPreferredColorScheme` to reuse the `usePreferredColorScheme` hook
- Update references to `withTheme` and `useStyle` in native code

## How has this been tested?
- Try the demo app on Android and iOS
- Try the demo app using Xcode 11 (beta) on iOS 13
- Set the simulator/device to DarkMode
- Check that all colors looks good and there are no errors.
    - All blocks
    - Bottom-sheets (Add link - Image settings)
    - Block inserter
    - Unsupported/Validation error placeholders
    - HTML mode

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by Tug** [comment]
@gziolo I'm addressing your feedback from https://github.com/WordPress/gutenberg/pull/17228 in this PR.

I did not yet add the `__unstable` or `__experimental` prefix as I don't believe this HOC/hook makes sense for the web at this stage and those native files are currently excluded for the documentation. If we want to port then we can always revisit the name, what do you think?

**Comment by gziolo** [comment]
> I did not yet add the `__unstable` or `__experimental` prefix as I don't believe this HOC/hook makes sense for the web at this stage and those native files are currently excluded for the documentation. If we want to port then we can always revisit the name, what do you think?

I think it's a broader issue of how we build expectations about the codebase. It's fine as of today, but as soon as we merge mobile and web repositories and fix all the missing declarations for dependencies, we should start treating all the public APIs equally for web and mobile. This means, we should be using `__unstable` (for APIs which needs polishing) and `__experimental` (for APIs which might get removed later).

File: packages/compose/src/index.native.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+// Web exports
+export * from './index.js';
```
Comment (gziolo) [inline packages/compose/src/index.native.js]: This one looks hacky but we can address it later by introducing two new files:

- `packages/compose/src/higher-order/index.js`
- `packages/compose/src/hooks/index.js`

**Review by gziolo — APPROVED** [review]
Code wise, this looks great. I haven't tested it. It'd would be great to confirm that it still works as expected. It looks like `useDarkModeContext` hook should do its job much better than the previous solution with the event emitter.

File: packages/block-editor/src/components/warning/index.native.js [hunk 2]
```diff
@@ -6,23 +6,24 @@ import { View, Text } from 'react-native';
...
 
 /**
  * Internal dependencies
  */
 import styles from './style.scss';
 
-function Warning( { title, message, icon, iconClass, theme, useStyle, ...viewProps } ) {
+function Warning( { title, message, icon, iconClass, theme, getStylesFromColorScheme, ...viewProps } ) {
 	icon = icon && normalizeIconObject( icon );
 	const internalIconClass = 'warning-icon' + '-' + theme;
```
Comment (etoledom) [inline packages/block-editor/src/components/warning/index.native.js]: `theme` here should probably be renamed to `preferredColorScheme`?
This should fix the Warning component icon color issue.

File: packages/block-library/src/missing/edit.native.js [hunk 3]
```diff
@@ -20,22 +21,22 @@ import styles from './style.scss';
...
 		const blockType = coreBlocks[ originalName ];
 
 		const title = blockType ? blockType.settings.title : __( 'Unsupported' );
-		const titleStyle = useStyle( styles.unsupportedBlockMessage, styles.unsupportedBlockMessageDark );
+		const titleStyle = getStylesFromColorScheme( styles.unsupportedBlockMessage, styles.unsupportedBlockMessageDark );
 
 		const icon = blockType ? normalizeIconObject( blockType.settings.icon ) : 'admin-plugins';
-		const iconStyle = useStyle( styles.unsupportedBlockIcon, styles.unsupportedBlockIconDark );
+		const iconStyle = getStylesFromColorScheme( styles.unsupportedBlockIcon, styles.unsupportedBlockIconDark );
 		const iconClassName = 'unsupported-icon' + '-' + theme;
```
Comment (etoledom) [inline packages/block-library/src/missing/edit.native.js]: Same than on Warning block, this looks like should be `preferredColorScheme`.

**Review by etoledom — COMMENTED** [review]
@Tug - This looks great! Thank you so much for these improvements 🎉 

I just spotted one very small detail:
- The icon on the Warning component should be white. Added a code comment with what I think is the problem.

dark_mode_v2

File: packages/block-editor/src/components/warning/index.native.js (hunk 2 above)
Comment (Tug) [reply packages/block-editor/src/components/warning/index.native.js]: Indeed, forgot that one thanks 👍 , will scan the code for other occurrences as well!

**Pull request merged**